### ISSUE TITLE
Ephemeral workers which re-register quickly will not be removed 

### ISSUE THREAD

**Issue by topherbullock** [concourse/concourse#2827]
## What challenge are you facing?

When using ephemeral workers (on the [Helm Chart](https://github.com/helm/charts/tree/master/stable/concourse)), if a worker dies and comes back quickly ( faster than GC will run the worker lifecycle collector ) it won't be removed from the db as intended.

## What would make this better?

Since we're moving more toward a world where workers are all registered as forwarded we could detect when the workers' forwarded connections die and handle the lifecycle on the TSA. Good thing we have a shiny new TSA client we can add features to 😉

**Comment by vito** [comment]
This is only an issue if the worker dies and comes back under the same name, right?

I haven't run the Helm chart lately so don't know how it picks the names, but it seems like if the container gets re-created it should get a new name. If a worker dies and comes back in-place quick enough, it kind of seems reasonable for it to not be removed, assuming the containers/volumes are all present and accounted for.

**Comment by topherbullock** [comment]
@vito Correct. Its only a problem if the container running the worker binary goes away and comes back up quickly ( before the worker lifecycle GC ). The chart uses the container name for the worker name (similar to Docker), so it might come back with the same name and different state.

This happens when the worker pod is rebalanced to another node and the volumes go away, but it is rebalanced quickly enough that it beats the worker lifecycle GC. Another way to frame the issue is that worker GC doesn't catch K8s' switch-a-roos of Pods fast enough.

**Comment by cirocosta** [comment]
Oh, regarding the naming, *as is right now*, it's "stable" like `${release}_worker_${n}` as @topherbullock described just because the chart relies on a StatefulSet (which guarantees this kind of stable naming) instead of a Deployment (which gives you some "random" names) (although I've been recently pushing more towards Deployments - hence those issues around making sure we don't leak things with Btrfs).

**Comment by vito** [comment]
Should we have `--ephemeral` workers just remove themselves on exit? :thinking: (Some context in https://github.com/concourse/concourse-docker/pull/44#issuecomment-491390754)
